### PR TITLE
Chelseag.profile 8450.customized samplers

### PR DESCRIPTION
**Pull request by chelseagzr**
## Motivation
The addition of two samplers `SingleTaskAdaptiveBatchSampler` and `SingleTaskSingleLengthAdaptiveBatchSampler` is driven by the need to address two critical bottlenecks that slow down multi-task training in our pipeline:
**Insufficient Batch Processing**:
When doing multi-task learning, the forward_loss function of the MultitaskModel class requires a separate forward pass for data points from each task. (reference: the [forward_loss](https://github.com/flairNLP/flair/blob/ee8596c2bbe737ec9ddeb1c6cb62fa0b161f4d84/flair/models/multitask_model.py#L69) function of MultitaskModel) For instance, if a batch contains data points from 3 different tasks, this results in 3 separate forward passes. Consequently, the memory usage and processing time for each batch can vary significantly based on the distribution of tasks within the batch. If one GPU is assigned a batch that takes longer to process, the remaining GPUs may become idle, leading to inefficient hardware utilization.
Solution: The first sampler, `SingleTaskAdaptiveBatchSampler`, ensures that each batch contains data points from a single task, thereby stabilizing memory usage and processing time across GPUs and improving overall efficiency.
**Padding Within Batches**:
When using SequenceTagger model and TextClassifier model, in a forward pass to process data points from the same task, each data point is padded to match the length of the longest data point in the batch before being input to the transformer layers. This padding introduces unnecessary computation and limits the batch size that can be used without causing CUDA out-of-memory errors, especially for batches containing long sequences.
Solution: The second sampler, `SingleTaskSingleLengthAdaptiveBatchSampler`, constructs batches where all data points in a batch not only come from the same task but also have the same length. This eliminates the need for padding, enabling the use of larger batch sizes for short sequences and smaller batch sizes for long sequences, further optimizing GPU memory usage and computational efficiency.

Additionally, to improve training efficiency, the samplers support adaptive downsampling of training data for each task based on a user-provided function. This is particularly useful when a certain performance metric (such as micro F1 score) has already reached a satisfactory level for a specific task. In such cases, we can reduce the number of training examples sampled from that task, thereby focusing computational resources on tasks that need more improvement and reducing overall training time.
For example, a user can specify a downsampling ratio such as:
downsample_ratio = 1 - (micro_f1 ** 2)
As the micro F1 score for a task increases, the downsampling ratio decreases, meaning fewer data points from that task are included in subsequent training batches.

## Changes
1. Added the new samplers in `flair/samplers.py`.
2. Added `fine_tune_with_customized_sampler` in `flair/trainers/trainer.py`.
    The `fine_tune_with_customized_sampler` function was added for the following purpose: for each optimizer update, gradients are computed using training data of varying lengths to help balance the gradients across each update step. To achieve this, the `num_batches_per_optimizer_step` parameter was introduced. For example, if using 4 GPUs and `num_batches_per_optimizer_step` is set to 16, each GPU processes 16 batches before each optimizer step, where each batch contains data points of a specific length. The lengths of datapoints in 16 batches are (l₁, l₂, ..., l₁₆). Importantly, the sequence of lengths (l₁, l₂, ..., l₁₆) is aligned across all GPUs—at any given batch index (e.g., the third batch on every GPU), all GPUs are processing data points of the same length. Before each optimizer step, gradients are accumulated over 64 batches in total (16 batches per GPU × 4 GPUs).

## Test
Ran a flyte workflow using SingleTaskSingleLengthAdaptiveBatchSampler on 5 tasks.

### CODE REVIEW

File: flair/trainers/trainer.py [hunk 1]
```diff
@@ -913,6 +916,588 @@ def wrapped_forward_loss(*args, **kwargs2):
 
         return return_values
 
+    def fine_tune_with_customized_sampler(
```
Comment (jeffpicard) [inline flair/trainers/trainer.py:919]: I think it would be better if the `sampler` param and its implementation were inside the existing `train_custom` method rather than duplicating the function, in order to avoid the implementations diverging. Do you have any idea how much work that would be, or if there are reasons not to do that?

File: flair/samplers.py [hunk 2]
```diff
@@ -114,3 +120,348 @@ def __iter__(self):
...
+
+        if shuffle:
+            # control randomness to ensure indices are identical across all GPUs
+            g = torch.Generator()
+            g.manual_seed(self.seed + epoch)
+            perm = torch.randperm(num_datapoints, generator=g).tolist()
+            indices = [indices[i] for i in perm]
+
+        # Distribute indices to each GPU
+        num_datapoints_per_GPU = (
```
Comment (ntravis22) [inline flair/samplers.py]: small nit: I'd stick with lowercase for variable names.

File: flair/samplers.py [hunk 3]
```diff
@@ -114,3 +120,348 @@ def __iter__(self):
...
+
+        Args:
+            batches (list[list[int]]): List of batches to shuffle.
+            epoch (int): Current epoch number for randomness.
+
+        Returns:
+            list[list[int]]: Shuffled list of batches.
+        """
+        # control randomness to ensure the permutation to shuffle batches are identical across all GPUs
+        g = torch.Generator()
```
Comment (ntravis22) [inline flair/samplers.py:219]: Looks like this logic is in the above

if shuffle:
    ...

clause. Do you want to call this method up there (in downsample_distribute_and_chunk_indices)?

File: flair/samplers.py [hunk 4]
```diff
@@ -114,3 +120,348 @@ def __iter__(self):
...
+        g.manual_seed(self.seed + epoch)
+        perm = torch.randperm(len(batches), generator=g).tolist()
+        return [batches[i] for i in perm]
+
+    def save_batches_to_s3(self, epoch: int) -> None:
```
Comment (ntravis22) [inline flair/samplers.py:224]: I get that this is useful for debugging but I wonder if we should remove it now that you are submitting the code. Wdyt?

**Review by jeffpicard — COMMENTED** [review]
This is so great!
- The new classes themselves LGTM! I just wonder if we can improve maintainability with relatively little work
- How much work do you think it would be to add a unit test on `generate` or something? (Not for the distributed case, just a small input dataset)

File: flair/samplers.py [hunk 5]
```diff
@@ -114,3 +120,348 @@ def __iter__(self):
...
+        """
+        if not isinstance(dataset, ConcatFlairDataset):
+            raise RuntimeError("SingleTaskAdaptiveBatchSampler only supports ConcatFlairDataset!")
+        assert len(dataset.cummulative_sizes) == len(dataset.ids)
+        # cummulative_sizes is an attribute of ConcatFlairDataset that defines the boundaries of each dataset within the concatenated dataset.
+        # For example, if cummulative_sizes = [5, 20, 50]:
+        # - Data points from the first task start at index 0 and end at index 5 (exclusive) in the ConcatFlairDataset.
+        # - Data points from the second task start at index 5 and end at index 20 (exclusive).
+        # - Data points from the third task start at index 20 and end at index 50 (exclusive).
+        self.cummulative_sizes = dataset.cummulative_sizes
```
Comment (ntravis22) [inline flair/samplers.py:405]: small nit: cummulative -> cumulative

**Review by ntravis22 — COMMENTED** [review]
This is excellent work @chelseagzr! Hats off!

+1 to @jeffpicard's comment. That would be my main concern also.

**Comment by chelseagzr** [comment]
> * How much work do you think it would be to add a unit test on `generate` or something? (Not for the distributed case, just a small input dataset)

I added two unit tests in `tests/test_samplers.py`. 
Note: I need to remove `--mypy` from `addopts = "--black --mypy --ruff"` in `flair/pyproject.toml` before running `pytest tests/test_samplers.py` because mypy detected some errors in files unrelated to this PR, e.g., flair/models/generative_classification_model.py.

File: flair/samplers.py (hunk 5 above)
Comment (chelseagzr) [reply flair/samplers.py:405]: This is a typo in flair: In `flair/flair/data.py`,
```
    @property
    def cummulative_sizes(self) -> list[int]:
        return self.cumulative_sizes
```
Since `cummulative_sizes` seems to be the interface to use, I'll just respect their typo.

File: flair/samplers.py (hunk 4 above)
Comment (chelseagzr) [reply flair/samplers.py:224]: We might still use this feature in the future, so I prefer keeping it.

File: flair/samplers.py (hunk 3 above)
Comment (chelseagzr) [reply flair/samplers.py:219]: Great question! The shuffle functionality needs to be preserved in both `downsample_distribute_and_chunk_indices` and `shuffle_batches` because they operate at different levels. Consider the case of `SingleTaskAdaptiveBatchSampler`, where each batch is guaranteed to contain data points from the same task. Suppose we have a concatenated dataset consisting of 4 sentences from task A and 4 from task B, like `[[a1, a2, a3, a4], [b1, b2, b3, b4]]` (where each `ai` or `bi` represents a sentence). Assume a batch size of 2. The shuffle in `downsample_distribute_and_chunk_indices` shuffles indices within each task, so `[a1, a2, a3, a4]` might become `[a1, a3, a4, a2]`. On the other hand, `shuffle_batches` shuffles the order of the batches themselves. For example, the batches `[[a1, a2], [a3, a4], [b1, b2], [b3, b4]]` could be rearranged to something like `[[a3, a4], [b3, b4], [b1, b2], [a1, a2]]`.

File: flair/trainers/trainer.py (hunk 1 above)
Comment (chelseagzr) [reply flair/trainers/trainer.py:919]: Great point! I think it would be super nice if I can merge `fine_tune_with_customized_sampler` into `train_custom` to help prevent future divergence. I did explore this approach, but it is challenging due to differences in handling `mini_batch_chunk_size` (used in `train_custom`) and `num_batches_per_optimizer_step` (used in `fine_tune_with_customized_sampler`). Also, I can’t simply remove `mini_batch_chunk_size` as an input because it’s used in several existing tests.

File: flair/samplers.py (hunk 3 above)
Comment (ntravis22) [reply flair/samplers.py:219]: Ok sounds good.

File: flair/samplers.py (hunk 4 above)
Comment (ntravis22) [reply flair/samplers.py:224]: Ok

File: flair/samplers.py (hunk 5 above)
Comment (ntravis22) [reply flair/samplers.py:405]: Gotcha, sounds good.

**Review by ntravis22 — APPROVED** [review]
Having to maintain a new copy of train_custom will not be ideal, but I guess that is a requirement for these speed improvements. Lgtm.

**Review by jeffpicard — APPROVED** [review]
Thanks for adding the tests!
